src/Unread.ts

```diff
@@ -59,36 +60,34 @@ export function doesRoomHaveUnreadMessages(room: Room): boolean {
         return false;
     }
 
+    for (const timeline of [room, ...room.getThreads()]) {
+        // If the current timeline has unread messages, we're done.
+        if (doesRoomOrThreadHaveUnreadMessages(timeline)) {
+            return true;
+        }
+    }
+    // If we got here then no timelines were found with unread messages.
+    return false;
+}
+
+export function doesRoomOrThreadHaveUnreadMessages(room: Room | Thread): boolean {
```

Comment:
(Mostly choosing a random place to start a thread.)

We need to figure out how to handle threads that exist before this change or users will see many rooms marked as unread as threads that exited before MSC3771 will likely not have a read receipt and will appear as unread.

Reply:
@gsouquet I think you were going to write a bit about our options here? From what I remember of our conversation we want to mark any "existing" threads as read up to their current point when this rolls out (so we don't mark all "old" threads as unread). Potential options:

1. Send a read receipt in each thread to the server and receive them down sync. (This likely wouldn't work since we would essentially DoS the server & we don't load all threads on load anyway.)
2. Mark any threads before a certain time as read. This isn't desirable since different people will start using the new code at different times.

I'm not sure we came up with others (besides something about changing the entire storage of events to be closer to what mobile does).

Reply:
We have discussed this during a web meeting last week. 
One alternative approach that has not been listed above is to figure out for each room when the first threaded read receipt has been sent and to discard all the threads that haven't received an update since that point in time from the computation made in this pull request

Sounds like the most sensible thing to do, we might need to force an initial sync, to make sure we have access to that data

Reply:
> when the first threaded read receipt has been sent and to discard all the threads that haven't received an update since that point in time from the computation made in this pull request

How do we compare "when" this way? Using the `ts` of the receipt and the `origin_server_ts` of the events?

Reply:
This work is now essentially in matrix-org/matrix-js-sdk#3020, although we might need more tweaks here.